logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/backup/RestoreBackupUseCase.kt

```diff
@@ -0,0 +1,167 @@
+import com.wire.kalium.logic.feature.backup.RestoreBackupResult.BackupRestoreFailure.DecryptionFailure
+import com.wire.kalium.logic.feature.backup.RestoreBackupResult.BackupRestoreFailure.IncompatibleBackup
+import com.wire.kalium.logic.feature.backup.RestoreBackupResult.BackupRestoreFailure.InvalidUserId
+import com.wire.kalium.logic.feature.backup.RestoreBackupResult.Failure
+import com.wire.kalium.logic.functional.Either
+import com.wire.kalium.logic.functional.fold
+import com.wire.kalium.logic.kaliumLogger
+import com.wire.kalium.logic.util.extractCompressedFile
+import com.wire.kalium.logic.wrapStorageRequest
+import com.wire.kalium.persistence.backup.DatabaseImporter
+import com.wire.kalium.util.KaliumDispatcher
+import com.wire.kalium.util.KaliumDispatcherImpl
+import kotlinx.coroutines.withContext
+import kotlinx.serialization.decodeFromString
+import kotlinx.serialization.json.Json
+import okio.Path
+import okio.Source
+import okio.buffer
+import okio.use
+
+interface RestoreBackupUseCase {
+
+    /**
+     * Restores a valid previously created backup file into the current database, respecting the current data if there is any overlap.
+     * @param extractedBackupRootPath The absolute file system path to the backup file.
+     * @param password the password used to encrypt the original backup file. Null if the file was not encrypted.
+     * @return A [RestoreBackupResult] indicating the success or failure of the operation.
+     */
+    suspend operator fun invoke(extractedBackupRootPath: Path, password: String?): RestoreBackupResult
+}
+
+internal class RestoreBackupUseCaseImpl(
+    private val databaseImporter: DatabaseImporter,
+    private val kaliumFileSystem: KaliumFileSystem,
+    private val userId: UserId,
+    private val dispatchers: KaliumDispatcher = KaliumDispatcherImpl,
+    private val idMapper: IdMapper = MapperProvider.idMapper()
+) : RestoreBackupUseCase {
+
+    override suspend operator fun invoke(extractedBackupRootPath: Path, password: String?): RestoreBackupResult =
+        withContext(dispatchers.io) {
+            runSanityChecks(extractedBackupRootPath, password).fold({ error ->
+                return@withContext error
+            }, { (encryptedFilePath, isPasswordProtected) ->
+                if (isPasswordProtected) {
+                    decryptExtractAndImportBackup(encryptedFilePath, extractedBackupRootPath, password!!)
+                } else {
+                    getDbPathAndImport(extractedBackupRootPath)
+                }
+            })
+        }
+
+    private suspend fun decryptExtractAndImportBackup(
+        encryptedFilePath: Path?,
+        extractedBackupRootPath: Path,
+        password: String
+    ): RestoreBackupResult {
+        val backupSource = kaliumFileSystem.source(encryptedFilePath!!)
+        val extractedBackupPath = extractedBackupRootPath / BACKUP_UNENCRYPTED_FILE_NAME
+        val backupSink = kaliumFileSystem.sink(extractedBackupPath)
```

Comment:
after the other place is fixed, this is failing due to extractedBackupPath pointing to a non-existing path